File: Go-Fiber/controllers/shoesController.go
package controllers

import (
	"context"
	"fmt"

	"github.com/fnhernandorena/golang210324/database"
	"github.com/fnhernandorena/golang210324/models"
	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

func GetShoes(c *fiber.Ctx) error {
	var shoes []models.Shoe
	results, err := database.Collection.Find(context.TODO(), bson.M{})
	if err != nil {
		return err
	}
	defer results.Close(context.TODO())

	for results.Next(context.TODO()) {
		var shoe models.Shoe
		err := results.Decode(&shoe)
		if err != nil {
			return err
		}
		shoes = append(shoes, shoe)
	}

	return c.JSON(&fiber.Map{
		"shoes": shoes,
	})
}

func CreateShoe(c *fiber.Ctx) error {
	var shoe models.Shoe
	c.BodyParser(&shoe)
	document := bson.D{
		{Key: "brand", Value: shoe.Brand},
		{Key: "model", Value: shoe.Model},
		{Key: "size", Value: shoe.Size},
	}
	result, err := database.Collection.InsertOne(context.TODO(), document)

	if err != nil {
		fmt.Println(err)
	}

	return c.JSON(&fiber.Map{
		"data": result,
	})
}
func GetShoe(c *fiber.Ctx) error {
	shoeID := c.Params("id")

	objID, err := primitive.ObjectIDFromHex(shoeID)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "Shoe ID is not valid",
		})
	}

	filter := bson.M{"_id": objID}
	var shoe models.Shoe
	err = database.Collection.FindOne(context.TODO(), filter).Decode(&shoe)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Error retrieving shoe",
		})
	}

	return c.JSON(shoe)
}

func DeleteShoe(c *fiber.Ctx) error {
	userID := c.Params("id")

	objID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "Shoe id not valid",
		})
	}

	filter := bson.M{"_id": objID}

	_, err = database.Collection.DeleteOne(context.TODO(), filter)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Error at delete shoe",
		})
	}

	return c.JSON(fiber.Map{
		"message": "Shoe deleted successfully",
	})
}
func UpdateShoe(c *fiber.Ctx) error {
	shoeID := c.Params("id")
	objID, err := primitive.ObjectIDFromHex(shoeID)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "Shoe ID is not a valid",
		})
	}
	var updatedShoe models.Shoe
	if err := c.BodyParser(&updatedShoe); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "Error at request shoe data",
		})
	}
	filter := bson.M{"_id": objID}
	update := bson.D{
		{Key: "brand", Value: updatedShoe.Brand},
		{Key: "model", Value: updatedShoe.Model},
		{Key: "size", Value: updatedShoe.Size},
	}
	ee, err := database.Collection.UpdateOne(context.TODO(), filter, update)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": "Error at update shoe",
		})
	}
	fmt.Printf("UpdateResult: %+v\n", ee)

	return c.JSON(fiber.Map{
		"message": "Shoe updated successfully",
	})
}

File: Python-Flask/src/controllers/ShoesController.py
from flask import request, Response
from bson import json_util, ObjectId
from pymongo import MongoClient

from repositories.ShoesRepository import coll
 
def create_shoe_controller():
    data = request.get_json()
    brand = data.get("brand", None)
    model = data.get("model", None)
    size = data.get("size", None)
    if brand:
      response = coll.insert_one({
        "brand": brand,
        "model": model,
        "size": size,
      })
      result = {
        "id": str(response.inserted_id),
        "brand": brand,
        "model": model,
        "size": size
      }
      return result
    else:
      return 'Invalid payload', 400
    
def get_shoes_controller():
    data = coll.find()
    result = json_util.dumps(data)
    return Response(result, mimetype='application/json')

def get_shoe_controller(id):
    data = coll.find_one({'_id': ObjectId(id)})
    result = json_util.dumps(data)
    return Response(result, mimetype='application/json')

def update_shoe_controller(id):
    data = request.get_json()
    if len(data) == 0:
      return 'Invalid payload', 400
    
    response = coll.update_one({'_id': ObjectId(id)}, {'$set': data})

    if response.modified_count >= 1:
        return 'Shoe updated successfully', 200
    else:
        return 'Shoe not found', 404

def delete_shoe_controller(id):
    response = coll.delete_one({'_id': ObjectId(id)})
    if response.deleted_count >= 1:
        return 'Shoe deleted successfully', 200
    else:
        return 'Shoe not found', 404